components/tidb_query_datatype/src/codec/mysql/vector.rs

```diff
@@ -147,20 +173,28 @@ impl<'a> VectorFloat32Ref<'a> {
         Ok(())
     }
 
+    fn index(&self, idx: usize) -> f32 {
+        let byte_index: usize = idx * 4;
+        if byte_index + 4 > self.value.len() {
+            panic!(
+                "Index out of bounds: index = {}, length = {}",
+                idx,
+                self.len()
+            );
+        }
+        let float_ptr = unsafe { self.value.as_ptr().add(byte_index) as *const f32 };
+        unsafe { float_ptr.read_unaligned() }
+    }
+
     pub fn l2_squared_distance(&self, b: VectorFloat32Ref<'a>) -> Result<f64> {
         self.check_dims(b)?;
         let mut distance: f32 = 0.0;
 
-        // Note: Must NOT calculate over va[i].value and vb[i].value directly (which is
-        // a &[NotNan]), because it will panic if the calculated result is NaN.
-        // See https://docs.rs/ordered-float/latest/ordered_float/struct.NotNan.html
-        let va = self.primitive_value();
-        let vb = b.primitive_value();
         for i in 0..self.len() {
-            // Hope this can be vectorized.
-            let diff = va[i] - vb[i];
+            let diff = self.index(i) - b.index(i);
```

Comment:
I'm worried this won't be vectorized because `index` function has a `if` branch. 
How about adding an `index_unchecked` version?

Reply:
Nice catch! This is indeed changed in a later PR: https://github.com/tidbcloud/cloud-storage-engine/pull/1468

Reply:
8% improvement in https://github.com/tidbcloud/cloud-storage-engine/pull/1468 seems too small between SIMD and non-SIMD so I suspect that the improvement just comes from reducing a branch and the rust compiler does not generate a SIMD version. Can we check the assembly to verify whether SIMD code is generated?

Reply:
According to @EricZequan 's previous benchmarks, actually compiler is already generating some kind of SIMD even when there are branches. The followings are the deasm for L1 distance:

```asm
                // /rustc/aedd173a2c086e558c2b66d3743b344f977621a7/library/core/src/intrinsics.rs : 2774
                copy_nonoverlapping(src, dst, count)
        movss xmm3, dword ptr [r15 + rdx]
        movss xmm0, dword ptr [r15 + rdx + 4]
                // /home/yang/Desktop/bounds-check-cookbook/src/bin/fibvec_naive_indexing.rs : 72
                let diff = x.index(i) - y.index(i);
        subss xmm3, dword ptr [rax + rdx]
                // /rustc/aedd173a2c086e558c2b66d3743b344f977621a7/library/std/src/f32.rs : 188
                unsafe { intrinsics::fabsf32(self) }
        andps xmm3, xmm1
                // /home/yang/Desktop/bounds-check-cookbook/src/bin/fibvec_naive_indexing.rs : 73
                distance += diff.abs();
        addss xmm3, xmm2
                // /home/yang/Desktop/bounds-check-cookbook/src/bin/fibvec_naive_indexing.rs : 72
                let diff = x.index(i) - y.index(i);
        subss xmm0, dword ptr [rax + rdx + 4]
                // /rustc/aedd173a2c086e558c2b66d3743b344f977621a7/library/std/src/f32.rs : 188
                unsafe { intrinsics::fabsf32(self) }
        andps xmm0, xmm1
                // /home/yang/Desktop/bounds-check-cookbook/src/bin/fibvec_naive_indexing.rs : 73
                distance += diff.abs();
        addss xmm0, xmm3
                // /rustc/aedd173a2c086e558c2b66d3743b344f977621a7/library/core/src/iter/range.rs : 729
                if self.start < self.end {
        add rdx, 8
        movaps xmm2, xmm0
        add rsi, -2
        jne .LBB6_20
.LBB6_12:
        test cl, 1
        je .LBB6_14
                // /rustc/aedd173a2c086e558c2b66d3743b344f977621a7/library/core/src/intrinsics.rs : 2774
                copy_nonoverlapping(src, dst, count)
        movss xmm1, dword ptr [r15 + rdx]
                // /home/yang/Desktop/bounds-check-cookbook/src/bin/fibvec_naive_indexing.rs : 72
                let diff = x.index(i) - y.index(i);
        subss xmm1, dword ptr [rax + rdx]
                // /rustc/aedd173a2c086e558c2b66d3743b344f977621a7/library/std/src/f32.rs : 188
                unsafe { intrinsics::fabsf32(self) }
        andps xmm1, xmmword ptr [rip + .LCPI6_0]
                // /home/yang/Desktop/bounds-check-cookbook/src/bin/fibvec_naive_indexing.rs : 73
                distance += diff.abs();
        addss xmm0, xmm1
.LBB6_14:
```

So 8% improvement only comes from reducing the branch check.

Note that the generated SIMD instructions are not most-efficient. We could use https://github.com/ashvardanian/SimSIMD as an improvement. (Not implemented though)

Reply:
Got it. Looking forward to further improvement by using https://github.com/ashvardanian/SimSIMD.